R/pcairPartition.R

```diff
@@ -9,40 +9,64 @@ pcairPartition <- function(kinobj, divobj,
-    kin.read <- kin.id.all %in% div.id.all
-    div.read <- div.id.all %in% kin.id.all
-    if(!(all(kin.read) & all(div.read))){
-        warning('kinobj and divobj contain non-matching samples; only partitioning those present in both objects')
+    if(is.null(kin.id.all)) {
+        stop('colnames must be provided for kinobj')
     }
 
     # filter to samples in sample.include
+    kin.read <- rep(TRUE, length(kin.id.all))
     if(!is.null(sample.include)){
-        if(!all(sample.include %in% kin.id.all) | !all(sample.include %in% div.id.all)){
-            warning('some samples in sample.include are not in kinobj or divobj; they will not be included')
+        if(!all(sample.include %in% kin.id.all)){
+            warning('some samples in sample.include are not in kinobj; they will not be included')
         }
 
         # subset the read indicators to samples in sample.include
-        kin.read <- kin.read & (kin.id.all %in% sample.include)
-        div.read <- div.read & (div.id.all %in% sample.include)
+        kin.read <- kin.id.all %in% sample.include
+    }
+
+    if (!is.null(divobj)) {
+        # sample.id from divergence matrix
+        div.id.all <- .readSampleId(divobj)
+
+        # check for ids provided
+        if(is.null(div.id.all)) {
+            stop('colnames must be provided for divobj')
+        }
+
+        # filter to samples in sample.include
+        div.read <- rep(TRUE, length(div.id.all))
+        if(!is.null(sample.include)){
+            if(!all(sample.include %in% div.id.all)){
+                warning('some samples in sample.include are not in divobj; they will not be included')
+            }
 
-        if(verbose) message('Working with ', sum(kin.read), ' samples')
+            # subset the read indicators to samples in sample.include
+            div.read <- div.id.all %in% sample.include
+        }
+        
+        # logical indicators of which samples are in both sets
+        kin.match <- kin.id.all %in% div.id.all
+        div.match <- div.id.all %in% kin.id.all
+        if(!(all(kin.match) & all(div.match))){
+            warning('kinobj and divobj contain non-matching samples; only partitioning those present in both objects')
+        }
+        kin.read <- kin.read & kin.match
+        div.read <- div.read & div.match
+        rm(kin.match, div.match)
+        id.both <- union(kin.id.all, div.id.all)
     }
+    
+    if(verbose) message('Working with ', sum(kin.read), ' samples')
 
     # checks on unrel.set
     if(!is.null(unrel.set)){
-        if(!all(unrel.set %in% kin.id.all) | !all(unrel.set %in% div.id.all)){
+        if(!all(unrel.set %in% id.both)){
             warning('some samples in unrel.set are not in kinobj or divobj; they will not be included')
             # subset unrel.set to only those in kinobj and divobj
-            unrel.set <- unrel.set[(unrel.set %in% kin.id.all) & (unrel.set %in% div.id.all)]
+            unrel.set <- unrel.set[(unrel.set %in% id.both)]
```

Comment:
@smgogarten I think before we were checking that unrel.set was in both kin.id.all and div.id.all.  This way I think it's checking that unrel.set is in *either* kin.id.all or div.id.all.  Right?

Reply:
You're right, I should have used `intersect` instead of `union` above. Also `id.both` is undefined if `divobj=NULL`, this should be `id.all` instead.